### ISSUE TITLE
Memory leak on windows

### ISSUE THREAD

**Issue by pomplesiegel** [r0x0r/pywebview#137]
There appears to be a memory leak on Windows, occurring whenever `webview.evaluate_js()` is called. I'm using pywebview v1.7  + "pythonnet" on Windows 10 64bit, with python3.6.2. The same issue does not exist on mac.

Code to reproduce  (most trivial example I could come up with)
```py
import webview
import threading
import time

threadStopSignal = threading.Event()
threadStopSignal.clear()

def checkOnDOM():
	global threadStopSignal 
	while threadStopSignal.is_set() == False:
		time.sleep(.01)
		webview.evaluate_js("doNothing()" )
		
def main():
	print("Beginning program")
	t = threading.Thread(target=checkOnDOM)
	t.start()

	webview.create_window("This is a window", "index.html")
	threadStopSignal.set()
	t.join(3.0)
main()
```

```html
<!DOCTYPE html>
<html>
<script type="text/javascript">
	function doNothing() { }
</script>
<head>
</head>
<body>
<div>
</div>
</body>
</html>
```

Then watch the memory usage of the program grow in the task manager. The memory usage on Mac remains static when running this program.

In order to further investigate, I later changed doNothing to 
```js
function doNothing() {
     return document.getElementsByTagName("script").length;
}
```
which, when printed in python, made it clear that the number of script elements is growing (and never shrinking) with each call to `evaluate_js()`. Not sure if that's the expected behavior. 

Finally, if the value of those script tags is printed, we see something like
`function invokecb2c8066aea211e7849010f0050b27cc() {return eval("doNothing()")}`
which over time monotonically grows to higher values 
```
function invokef9d07746aea211e780ee10f0050b27cc() {return eval("doNothing()")}
function invokef9d29312aea211e791b410f0050b27cc() {return eval("doNothing()")}
function invokef9d46ef6aea211e78f1f10f0050b27cc() {return eval("doNothing()")}
```
up and up.... Are these pointers to resources in windows that never get cleaned up, or are they just GUIDs?

Happy to provide any more info. 

Really enjoying the library, btw! Hoping to resolve the memory issue on the windows side so I can use it further.

**Comment by denfromufa** [comment]
We noticed a memory leak in pythonnet and it is fixed in pull request for most use cases:

https://github.com/pythonnet/pythonnet/issues/541

**Comment by shivaprsdv** [comment]
@pomplesiegel This is interesting. 🤔  Everything you have reported on
`evaluate_js()` is expected behaviour (except for the memory leak, of course).
The script elements and UUIDs are part of the implementation, you can see it in
the source. They were added to return the value of the executed JavaScript.

I don't reckon this should has anything to do with the leak, after all, does adding
tiny script elements to a page tend to blow up memory usage on normal browsers?
Have you ever noticed similar behaviour with any of the other API methods?

If this is correlated then we could delete the script elements after its use. Roman
wrote the code for WinForms, may be he can look into the issue.

**Comment by shivaprsdv** [comment]
@pomplesiegel This is interesting. 🤔  Everything you have reported on
`evaluate_js()` is expected behaviour (except for the memory leak, of course).
The script elements and UUIDs are part of the implementation, you can see it in
the [source](https://github.com/r0x0r/pywebview/blob/03b763b9bbb5df8316ee622c7d3f1eebdc5e640f/webview/winforms.py#L251).

Before that, your program is doing an infinite loop. I reckon you are doing
this on purpose? Because, ` threadStopSignal.set()` will never get called
because the call to `create_window` blocks the main thread.

I think we have indeed hit a bug here. Not deleting the script elements
after use could be the cause. Roman wrote the code for WinForms, may
be he'll look into this, pronto.

**Comment by pomplesiegel** [comment]
@shivaprsdv Great, yes I believe this is a bug as well. Thank you. 

Off topic: You're right about the infinite loop in this example code. How would you recommend calling `destroy_window()` when the user hits the "x" button on the GUI window?

**Comment by r0x0r** [comment]
You do not need to call destroy_window to close the window by pressing the X button. This case is already handled by webview. As Shiva said Invoke eval stuff is implemented on purpose to get a return value from the evaluated code.

**Comment by pomplesiegel** [comment]
Hi @r0x0r, I apologize for getting off topic. I removed the second half of my post. 
This issue is focused on the memory leak in windows caused by repeatedly calling `evaluate_js()`. With the program above, the memory usage on my Windows machine will grow by multiple MB per minute. Do you see this behavior on your machine as well?

**Comment by shivaprsdv** [comment]
The leak should be real. Although it doesn't happen on Mac as the OP pointed
out, I tried simulating what must be happening; it ate all my RAM, and in the Web
Inspector I found a heap growth:
##### Normal on Mac:
![normal](https://user-images.githubusercontent.com/18247990/31502040-8fac1ece-af89-11e7-9764-11c41df351b7.png)
##### Simulated:
![simulated](https://user-images.githubusercontent.com/18247990/31502277-37741e68-af8a-11e7-8c22-789472453bfa.png)

The issue is simple: every time we call `evaluate_js()`, we add a script element
to the doc, and a function to the JS context. The functions would not be garbage
collected, so when we call it a thousand times it blows up the memory.

Leak or not, ideally we should be deleting the script element after it accomplishes
its purpose anyway; it's not good to permanently modify the user's page like this.

**Comment by r0x0r** [comment]
I see the problem now, but this is how the Windows implementation works. The problem is that Windows does not provide a way to run arbitrary Javascript, but instead is able to run a named function. So in order to run Javascript code, it must be wrapped into a named function, which will be called (wrapped inside an eval to get its return value). Given all that, the function will stay in memory.
I guess we could fit a del after the function call. I will investigate if it is a feasible way and how it will play with async code.

**Comment by pomplesiegel** [comment]
Great, thank you!

**Comment by shivaprsdv** [comment]
@r0x0r Could add the code to delete the function along with the `script` element
once `_js_result_semaphor()` releases.

**Comment by r0x0r** [comment]
Gave it a go and found out the following 
1) `delete` works only on object properties, not top level objects or functions
2) You cannot delete a <script> element using removeNode or removeChild

The next step will be encapsulating an invoke function inside an object and see if WinForms is able to call that. It is late here though, I will give it a go at a more appropriate time.

**Comment by r0x0r** [comment]
Uhm the object property approach is no good either. Any other ideas?
![capture](https://user-images.githubusercontent.com/1549657/31520195-bde0dbbc-afad-11e7-9f91-6a1a29c74a90.PNG)

**Comment by shivaprsdv** [comment]
1) JavaScript is garbage collected, so no need to use `delete`; set the function to
`null` instead.
2) Set the `OuterHTML` property of the <script> to empty string.

I would try:
```py
    ...
    code = 'function {0}() {{ {0} = null; return eval("{1}"); }}'  # 1.
    c0de = code.format(function_name, _escape_string(script))
    ...
    self._js_result = document.InvokeScript(function_name)
    self._js_result_semaphor.release()
    script_element.OuterHTML = ""  # 2.
    ...

**Comment by shivaprsdv** [comment]
I have pushed ab355cc956d5a80a8e99b370c953c8e822f0baf3 with the above modifications.

@pomplesiegel Can you check whether this solves the problem? I have no way to test it.

**Comment by pomplesiegel** [comment]
Hi @shivaprsdv, thanks for the update! I just checked out and ran this branch on windows and unfortunately the memory leak still persists on Windows. I'm happy to try anything further.

**Comment by shivaprsdv** [comment]
I must have overlooked something. Have you tried printing the script tags this
time, like you did earlier? Can you check whether it still grows?

**Comment by pomplesiegel** [comment]
Sure, i just ran it w/ the program printing a return value of 
```js
return document.getElementsByTagName("script").length;
```
and this value is growing rapidly. 

If I print the last value of the script array as it grows, i get something like
```js
function invokee3399462b44a11e7a86110f0050b27cc() { invokee3399462b44a11e7a86110f0050b27cc = null; return eval("doNothing()"); }
function invokee33b6676b44a11e78de910f0050b27cc() { invokee33b6676b44a11e78de910f0050b27cc = null; return eval("doNothing()"); }
function invokee33d7938b44a11e7af7010f0050b27cc() { invokee33d7938b44a11e7af7010f0050b27cc = null; return eval("doNothing()"); }
function invokee33f8988b44a11e7805810f0050b27cc() { invokee33f8988b44a11e7805810f0050b27cc = null; return eval("doNothing()"); }
function invokee341662eb44a11e7852210f0050b27cc() { invokee341662eb44a11e7852210f0050b27cc = null; return eval("doNothing()"); }
function invokee34348dab44a11e7a62610f0050b27cc() { invokee34348dab44a11e7a62610f0050b27cc = null; return eval("doNothing()"); }
function invokee3456474b44a11e79fde10f0050b27cc() { invokee3456474b44a11e79fde10f0050b27cc = null; return eval("doNothing()"); }
function invokee34773e6b44a11e79c7210f0050b27cc() { invokee34773e6b44a11e79c7210f0050b27cc = null; return eval("doNothing()"); }
```

**Comment by shivaprsdv** [comment]
@r0x0r Do we need all this? The guy at [this page](https://www.codeproject.com/Tips/60924/Using-WebBrowser-Document-InvokeScript-to-mess-aro) says:

![dsdf](https://user-images.githubusercontent.com/18247990/31775397-25042292-b506-11e7-9b10-2595af886748.png)

I remember the talk we had on this on the original evaluate-js issue; nevertheless,
had you tried this before deciding to settle on the `invoke3399...` idea?

**Comment by r0x0r** [comment]
Good find. I haven't thought of using eval directly, I will give it a go.

**Comment by r0x0r** [comment]
I pushed a revised version to the master. For some reason it refuses to work with the escaped script, but the original script works like a charm (with line breaks and line comments)
@pomplesiegel Could you verify that it does not leak memory?

**Comment by pomplesiegel** [comment]
Hi @r0x0r, thanks for the update!

This is an improvement, as the size of the scripts is no longer increasing. However I do still see a fairly fast leak with my empty program from above: About 1MB every 20s.

Do you see this on your side as well using Task Manager?

**Comment by shivaprsdv** [comment]
> but the original script works like a charm

@r0x0r Have you checked there is no need to escape even _double quotes_? 😃 

> I do still see a fairly fast leak with my empty program from above: About 1MB
every 20s.

@pomplesiegel May be this is normal, seeing you're running an intensive infinite
loop. To pin point the issue, can you call any other API function (say `load_html`) 
instead of `evaluate_js`, and compare the memory usage?

**Comment by pomplesiegel** [comment]
Hi @shivaprsdv, i just checked this. The program does not grow when calling `load_html("")`. Also, on the Mac side the program does not grow when calling `evaluate_js()`, so this issue appears to be linked to just `evaluate_js()` and just on Windows.

**Comment by shivaprsdv** [comment]
Is there any way to check the behaviour of this program (or something similar) on
Internet Explorer? It could well be an MSHTML issue, for all I can say.. 🤔

**Comment by r0x0r** [comment]
Here is what I found out. 
1) As it was pointed above, your original program has got an infinite loop. Remember that create_window blocks execution, until the window is destroyed
2) I have modified the script to invoke doNothing repeatedly without invoking evaluate_js and memory grows in the same fashion as with evaluate_js. Furthermore IE does the same.
```
<!DOCTYPE html>
<html>
<script type="text/javascript">
	function doNothing() { }

	for(var i =0; i < 100000; i++) {
	    doNothing();
    }
</script>
<head>
</head>
<body>
<div>
    <button onclick="doNothing()">Test</button>
</div>
</body>
</html>
```

3) All in all, it seems this is how IE garbage collection works and there is nothing we can do about that. A good thing that we got rid script elements and invoke functions, though!
